### PR TITLE
e2e, ACPI: Add NoACPI decorator

### LINKED ISSUE #13501: arm64: failing test `VMIlifecycle Softreboot a VirtualMachineInstance soft reboot vmi with agent connected should succeed`
# What happened

Failing test detected: `VMIlifecycle Softreboot a VirtualMachineInstance soft reboot vmi with agent connected should succeed` [1]
Example failure [2]

/kind failing-test
/priority critical-urgent

/sig compute
/wg arch-arm

/assign @zhlhahaha 

[1]: https://search.ci.kubevirt.io/?search=VMIlifecycle+Softreboot+a+VirtualMachineInstance+soft+reboot+vmi+with+agent+connected+should+succeed&maxAge=336h&context=1&type=bug%2Bissue%2Bjunit&name=&excludeName=&maxMatches=5&maxBytes=20971520&groupBy=job
[2]: https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/directory/pull-kubevirt-e2e-arm64/1867028566642790400

## Additional context

# Flake Action Plan

**As the assignee**, **thoroughly review the issue** and put the resulting report as comment into this issue.

### **The flake is a bug**

* Add label commenting `/triage accepted`
* **Create a PR to fix the bug**
* Reference this issue in the PR
* Keep this issue open until the testcase does not fail anymore

### **The flake is non-critical or an issue that is hard to fix**
* **Quarantine the test** by creating a pull request assigning [QUARANTINE] tag to test name and [Quarantine decorator](https://github.com/kubevirt/kubevirt/blob/f85a8117b8a90fd913ec0719faae9506866d1525/tests/decorators/decorators.go#L7) to the test
* Reference this issue on the PR

### There was an infrastructure issue

An infra issue is anything "below" the testcase

* Add label commenting `/triage infra-issue`
* Close the issue, adding a comment with details about the infrastructure issue.

After the flake has been fixed, document the learning from it inside the fix PR

### PR DESCRIPTION
**Pull request by zhlhahaha**
ACPI feature is needed on Arm64 for UEFI boot.

Fixes #13501

release note
```release-note
NONE
```

### CODE REVIEW

**Comment by zhlhahaha** [comment]
cc: @dhiller @brianmcarey

**Comment by iholder101** [comment]
Hey @zhlhahaha! Thank you for this PR!

We recently had efforts to remove skips from tests and instead use decorators that would be able to be used in order to skip these tests. See https://github.com/kubevirt/kubevirt/blob/main/tests/decorators/decorators.go.

Perhaps we can add another label under the `HW` section instead?

**Comment by iholder101** [comment]
Turns out this function is already marked as deprecated:
https://github.com/kubevirt/kubevirt/blob/94645722e3a62a9a1bc15e734d322437650101c0/tests/framework/checks/skips.go#L235-L236

/hold

**Comment by zhlhahaha** [comment]
> // Deprecated: SkipIfARM64 should be converted to check & fail

Sorry, I did not notice the comments. Thanks for reminding.

> Perhaps we can add another label under the HW section instead?

Do you means ACPI? I think it's supposed to be in the `// Features` section.

> We recently had efforts to remove skips from tests and instead use decorators that would be able to be used in order to skip these tests. See https://github.com/kubevirt/kubevirt/blob/main/tests/decorators/decorators.go.

I’m not very familiar with decorators, but it seems that they are similar to the e2e test tags/labels. For example, `decorators.SigCompute` appears to correspond to `[sig-compute]` based on the definitions in decorators.go.

How can we skip ACPI-related tests on Arm64?
1. Add an ACPI decorator in decorators.go, e.g., `ACPI = Label("ACPI")`
2. Add the ACPI decorator to the relevant tests, e.g.:
```
It("soft reboot vmi with agent connected should succeed", decorators.Conformance, decorators.ACPI, func() {...})
```
3. Skip these tests by setting `KUBEVIRT_E2E_SKIP`

Is my understanding correct?

File: tests/vmi_lifecycle_test.go [hunk 1]
```diff
@@ -1572,6 +1572,7 @@ var _ = Describe("[rfe_id:273][crit:high][arm64][vendor:cnv-qe@redhat.com][level
 		const vmiLaunchTimeout = 360
 
 		It("soft reboot vmi with agent connected should succeed", decorators.Conformance, func() {
+			checks.SkipIfARM64(testsuite.Arch, "arm64 need ACPI to boot for UEFI boot")
```
Comment (0xFelix) [inline tests/vmi_lifecycle_test.go:1575]: Not sure this is using UEFI at all btw?

**Comment by iholder101** [comment]
> > // Deprecated: SkipIfARM64 should be converted to check & fail
> 
> Sorry, I did not notice the comments. Thanks for reminding.
> 
> > Perhaps we can add another label under the HW section instead?
> 
> Do you means ACPI? I think it's supposed to be in the `// Features` section.

Also works :+1: 

> > We recently had efforts to remove skips from tests and instead use decorators that would be able to be used in order to skip these tests. See [`main`/tests/decorators/decorators.go](https://github.com/kubevirt/kubevirt/blob/main/tests/decorators/decorators.go?rgh-link-date=2024-12-16T12%3A17%3A36Z).
> 
> I’m not very familiar with decorators, but it seems that they are similar to the e2e test tags/labels. For example, `decorators.SigCompute` appears to correspond to `[sig-compute]` based on the definitions in decorators.go.
> 
> How can we skip ACPI-related tests on Arm64?
> 
> 1. Add an ACPI decorator in decorators.go, e.g., `ACPI = Label("ACPI")`
> 2. Add the ACPI decorator to the relevant tests, e.g.:
> 
> ```
> It("soft reboot vmi with agent connected should succeed", decorators.Conformance, decorators.ACPI, func() {...})
> ```
> 
> 3. Skip these tests by setting `KUBEVIRT_E2E_SKIP`
> 
> Is my understanding correct?

Yes it is! :)

File: tests/vmi_lifecycle_test.go (hunk 1 above)
Comment (zhlhahaha) [reply tests/vmi_lifecycle_test.go:1575]: > Not sure this is using UEFI at all btw?

Yes, currently all Virtual Machine on Arm64 platform using UEFI boot.

**Comment by zhlhahaha** [comment]
> > 3. Skip these tests by setting `KUBEVIRT_E2E_SKIP`
> > 
> > Is my understanding correct?
> 
> Yes it is! :)

Hi @iholder101 , I gave it a try to skip the tests via decorators and `KUBEVIRT_E2E_SKIP` but failed. Here is what I do:
1. Add `NoACPI` decorator, and add it to the specific tests
2. To verify the test I add a temporary label `[arm_test]` to the parent test suits.
3. Set `KUBEVIRT_E2E_SKIP=NoACPI` and `KUBEVIRT_E2E_FOCUS=arm_test`

However, the test cases with `decorators.NoACPI` are not skipped. I also tried to add label `[NoACPI]` directly on the test cases, and it works (test cases can be successfully skipped). Are there something wrong with my code? 
```
diff --git a/tests/decorators/decorators.go b/tests/decorators/decorators.go
index d3c3cd0aa9..c59ceec6ae 100644
--- a/tests/decorators/decorators.go
+++ b/tests/decorators/decorators.go
@@ -52,6 +52,7 @@ var (
        RequiresTwoWorkerNodesWithCPUManager = Label("requires-two-worker-nodes-with-cpu-manager")
        RequiresDualStackCluster             = Label("requires-dual-stack-cluster")
        RequiresHugepages2Mi                 = Label("requireHugepages2Mi")
+       NoACPI                               = Label("NoACPI")

        // Storage classes
        // Requires a storage class with support for snapshots
diff --git a/tests/vmi_lifecycle_test.go b/tests/vmi_lifecycle_test.go
index b3fafc57c2..c8b1f35fc2 100644
--- a/tests/vmi_lifecycle_test.go
+++ b/tests/vmi_lifecycle_test.go
@@ -1568,11 +1568,10 @@ var _ = Describe("[rfe_id:273][crit:high][arm64][vendor:cnv-qe@redhat.com][level
                })
        })

-       Describe("Softreboot a VirtualMachineInstance", func() {
+       Describe("[arm_test]Softreboot a VirtualMachineInstance", func() {
                const vmiLaunchTimeout = 360

-               It("soft reboot vmi with agent connected should succeed", decorators.Conformance, func() {
-                       checks.SkipIfARM64(testsuite.Arch, "arm64 need ACPI to boot for UEFI boot")
+               It("soft reboot vmi with agent connected should succeed", decorators.Conformance, decorators.NoACPI, func() {
                        vmi := libvmops.RunVMIAndExpectLaunch(libvmifact.NewFedora(withoutACPI()), vmiLaunchTimeout)

                        Eventually(matcher.ThisVMI(vmi), 12*time.Minute, 2*time.Second).Should(matcher.HaveConditionTrue(v1.VirtualMachineInstanceAgentConnected))
@@ -1595,8 +1594,7 @@ var _ = Describe("[rfe_id:273][crit:high][arm64][vendor:cnv-qe@redhat.com][level
                        waitForVMIRebooted(vmi, console.LoginToCirros)
                })

-               It("soft reboot vmi neither have the agent connected nor the ACPI feature enabled should fail", decorators.Conformance, func() {
-                       checks.SkipIfARM64(testsuite.Arch, "arm64 need ACPI to boot for UEFI boot")
+               It("soft reboot vmi neither have the agent connected nor the ACPI feature enabled should fail", decorators.Conformance, decorators.NoACPI, func() {
                        vmi := libvmops.RunVMIAndExpectLaunch(libvmifact.NewCirros(withoutACPI()), vmiLaunchTimeout)

                        Expect(console.LoginToCirros(vmi)).To(Succeed())
```

**Comment by iholder101** [comment]
@zhlhahaha The way you applied the `[arm_test]` is the old way of doing it without an official ginkgo label. Please try the following.

Add a new arm test label:
```go
NoACPI = Label("NoACPI")
Arm    = Label("[arm_test]")
```

Please try to replace:
```go
Describe("[arm_test]Softreboot a VirtualMachineInstance", func() {
```

With
```go
Describe("Softreboot a VirtualMachineInstance", decorators.Arm, func() {
```

This is equivalent to:
```go
Describe("[arm_test]Softreboot a VirtualMachineInstance", Label("[arm_test]"), func() {
```

(disclaimer: this is just an example, not a concrete suggestion to add these exact labels to the codebase).

**Comment by zhlhahaha** [comment]
> @zhlhahaha The way you applied the `[arm_test]` is the old way of doing it without an official ginkgo label. Please try the following.

Yes, I agree with you. This tag is only for testing purpose and will remove in the formal commit, so I add it in this way for convenience.
I want to test if we can successfully skip the NoACPI cases by using KUBEVIRT_E2E_SKIP, but it seems not work. Or did I do it in a wrong way?

**Comment by iholder101** [comment]
> > @zhlhahaha The way you applied the `[arm_test]` is the old way of doing it without an official ginkgo label. Please try the following.
> 
> Yes, I agree with you. This tag is only for testing purpose and will remove in the formal commit, so I add it in this way for convenience. I want to test if we can successfully skip the NoACPI cases by using KUBEVIRT_E2E_SKIP, but it seems not work. Or did I do it in a wrong way?
> 
> After setting the `KUBEVIRT_E2E_SKIP` and `KUBEVIRT_E2E_FOCUS `, I use `make functest` to run the e2e tests.

You seem to do it right.
Here's another [example](https://github.com/kubevirt/project-infra/blob/e930ebfdd8f7898e5e2d91a3270abbba43b22f8d/github/ci/prow-deploy/files/jobs/kubevirt/kubevirt/kubevirt-presubmits-0.42.yaml#L321-L322) for using it:
```yaml
        - name: KUBEVIRT_E2E_SKIP
          value: Multus|SRIOV|GPU|Macvtap|\[sig-operator\]
```

Maybe you need to escape the square brackets as shown here. But in essence, setting this variable (via `export KUBEVIRT_E2E_SKIP="something"`) should work.

You can also find some additional documentation here: https://github.com/kubevirt/kubevirt/blob/main/tests/README.md#run-them-on-one-of-the-core-kubevirt-providers.

**Comment by zhlhahaha** [comment]
> Maybe you need to escape the square brackets as shown here. But in essence, setting this variable (via `export KUBEVIRT_E2E_SKIP="something"`) should work.

Yes, you are right, after I set `export KUBEVIRT_E2E_SKIP="\[NoACPI\]"`, the test cases can be successfully skipped.

**Comment by dhiller** [comment]
IIUC currently the only config for `e2e-arm` is that tests being executed should have the arm label
* https://github.com/kubevirt/project-infra/blob/main/github/ci/prow-deploy/files/jobs/kubevirt/kubevirt/kubevirt-periodics.yaml#L770
* https://github.com/kubevirt/project-infra/blob/main/github/ci/prow-deploy/files/jobs/kubevirt/kubevirt/kubevirt-presubmits.yaml#L1022
So my question is: how does the `NoACPI` label get applied?

**Comment by iholder101** [comment]
> @iholder101 I was expecting a modification in automation/test.sh where the lane for arm would be configured correctly. Can you point me towards the configuration for the lane I have missed? Or is that done in the job configuration? If so I have missed that also.
>
> Thank you for pointing me towards the right direction.

Ehum.. in `automation/test.sh` I don't see that skips are being configured for lanes. So I think the job configuration is the right place.

> So my question is: how does the NoACPI label get applied?

This label doesn't currently exist, we were talking about such a theoretical label that maybe needs to be added.
We can perhaps add a "positive" `ACPI` label and skip it for non-x86 architectures.
Another option is to set up an `x86` label and skip it for all other archs.

File: tests/vmi_lifecycle_test.go (hunk 1 above)
Comment (0xFelix) [reply tests/vmi_lifecycle_test.go]: Okay, that makes senses. Sorry for the noise.

**Comment by dhiller** [comment]
> > @iholder101 I was expecting a modification in automation/test.sh where the lane for arm would be configured correctly. Can you point me towards the configuration for the lane I have missed? Or is that done in the job configuration? If so I have missed that also.
> > Thank you for pointing me towards the right direction.
> 
> Ehum.. in `automation/test.sh` I don't see that skips are being configured for lanes. So I think the job configuration is the right place.

To be more specific: I'd have expected an addition in [this section of automation/test.sh](https://github.com/kubevirt/kubevirt/blob/dd399cd9d45628c0e3758ff98e0d5d1576c29387/automation/test.sh#L405-L480). I believe that would be the place where config related to arm lanes should go - am I wrong?

> 
> > So my question is: how does the NoACPI label get applied?
> 
> This label doesn't currently exist, we were talking about such a theoretical label that maybe needs to be added. We can perhaps add a "positive" `ACPI` label and skip it for non-x86 architectures. Another option is to set up an `x86` label and skip it for all other archs.

**Comment by iholder101** [comment]
> > > @iholder101 I was expecting a modification in automation/test.sh where the lane for arm would be configured correctly. Can you point me towards the configuration for the lane I have missed? Or is that done in the job configuration? If so I have missed that also.
> > > Thank you for pointing me towards the right direction.
> > 
> > 
> > Ehum.. in `automation/test.sh` I don't see that skips are being configured for lanes. So I think the job configuration is the right place.
> 
> To be more specific: I'd have expected an addition in [this section of automation/test.sh](https://github.com/kubevirt/kubevirt/blob/dd399cd9d45628c0e3758ff98e0d5d1576c29387/automation/test.sh#L405-L480). I believe that would be the place where config related to arm lanes should go - am I wrong?
> 
> > > So my question is: how does the NoACPI label get applied?
> > 
> > 
> > This label doesn't currently exist, we were talking about such a theoretical label that maybe needs to be added. We can perhaps add a "positive" `ACPI` label and skip it for non-x86 architectures. Another option is to set up an `x86` label and skip it for all other archs.

Oh, got it! Thanks for clarifying.
Looks good to me :)

**Comment by zhlhahaha** [comment]
Hi, @dhiller @iholder101 
I am not sure what I supposed to do in this PR.

**Comment by iholder101** [comment]
> Hi, @dhiller @iholder101 I am not sure what I supposed to do in this PR.

Here's what I suggest:
1. Add an `ACPI` decorator / label under the `HW` section in `decorators.go`.
2. Add the label to tests like `It("soft reboot vmi with ACPI feature enabled should succeed"`.
3. Add something similar to [this section of automation/test.sh](https://github.com/kubevirt/kubevirt/blob/dd399cd9d45628c0e3758ff98e0d5d1576c29387/automation/test.sh#L405-L480):
```shell
elif [[ $TARGET =~ arm64 ]]; then
    label_filter='!(ACPI)'
```

@dhiller, WDYT? Please review my suggestion.

(p.s. as a follow-up, we might want to think of a way to easily skip relevant tests per-arch in order to run them easily locally.)

**Comment by zhlhahaha** [comment]
Hi @dhiller Do you have any comments or suggestions?

**Comment by zhlhahaha** [comment]
Amend this PR after the https://github.com/kubevirt/project-infra/pull/3850 get merged

**Comment by zhlhahaha** [comment]
Merge this PR to https://github.com/kubevirt/kubevirt/pull/13578

**Pull request closed without merging**